File: 2017/14/disk-defragmentation.py
# -*- coding: utf-8 -*-
import sys
import functools
import operator


SUFFIX = [17, 31, 73, 47, 23]


def build_hash(text, rounds=64):
    lengths = text.encode('utf-8') + bytes(SUFFIX)
    marks = list(range(256))
    skip = index = 0
    for _ in range(rounds):
        for length in lengths:
            twist(marks, index, length)
            index += length + skip
            index %= len(marks)
            skip += 1
    return marks


def reduce_hash(sparse_hash):
    dense_hash = []
    for b in range(0, 256, 16):
        block = sparse_hash[b:b + 16]
        compressed_block = functools.reduce(operator.xor, block)
        dense_hash.append(compressed_block)
    return dense_hash


def twist(marks, index, length):
    stack = []
    for i in range(length):
        stack.append(marks[(index + i) % 256])
    for i in range(length):
        marks[(index + i) % 256] = stack.pop(-1)


def get_bin(dense_hash):
    return ''.join('{:016b}'.format(b) for b in dense_hash)


# # #


def get_disk_state(key_string):
    grid = []
    for i in range(128):
        row_key = f'{key_string}-{i}'
        knot_hash = get_knot_hash(row_key)
        row_state = get_row_state(knot_hash)
        grid.append(row_state)
    return grid


def get_knot_hash(text):
    sparse_hash = build_hash(text)
    dense_hash = reduce_hash(sparse_hash)
    return dense_hash


def get_row_state(hash_):
    binary = get_bin(hash_)
    return [bit == '1' for bit in binary]


def count_used(grid):
    return sum(sum(row) for row in grid)


def main(key_string):
    grid = get_disk_state(key_string)
    num_used = count_used(grid)
    print(num_used)


if __name__ == '__main__':
    main(sys.argv[1])

File: 2017/14/disk-defragmentation.go
package main

import "fmt"
import "os"


const NUM_KNOTS = 256
const NUM_KEYS = 128
const NUM_ROUNDS = 64


type SparseHash = [NUM_KNOTS]byte
type Coord [2]int


var SUFFIX = []byte{17, 31, 73, 47, 23}


type ByteNode struct {
    Value byte
    Next *ByteNode
}

type ByteStack struct {
    Top *ByteNode
}

func (s *ByteStack) Push(value byte) {
    node := &ByteNode{Value: value, Next: s.Top}
    s.Top = node
}

func (s *ByteStack) Pop() byte {
    node := s.Top
    s.Top = s.Top.Next
    return node.Value
}


////


func buildSparseHash(lengths []byte) SparseHash {
    var marks SparseHash
    for i := 0; i < NUM_KNOTS; i += 1 {
        marks[i] = byte(i)
    }
    var index byte
    var skip int
    for r := 0; r < NUM_ROUNDS; r += 1 {
        for _, length := range lengths {
            twist(&marks, index, length)
            index += byte(int(length) + skip)
            skip += 1
        }
    }
    return marks
}


func twist(marks *SparseHash, index byte, length byte) {
    var segment ByteStack
    var i byte
    for i = 0; i < length; i += 1 {
        segment.Push(marks[index + i])
    }
    for i = 0; i < length; i += 1 {
        marks[index + i] = segment.Pop()
    }
}


func reduceHash(sparseHash SparseHash) (denseHash []byte) {
    var value byte
    for i, knot := range sparseHash {
        value ^= knot
        if (i + 1) % 16 == 0 {
            denseHash = append(denseHash, value)
            value = 0
        }
    }
    return
}


func getKnotHash(text string) []byte {
    var lengths = append([]byte(text), SUFFIX...)
    var sparseHash = buildSparseHash(lengths)
    var denseHash = reduceHash(sparseHash)
    return denseHash
}


////


func getDiskState(keyString string) map[Coord]bool {
    var grid = make(map[Coord]bool, 0)
    for y := 0; y < NUM_KEYS; y++ {
        var rowKey = fmt.Sprintf("%s-%d", keyString, y)
        var knotHash = getKnotHash(rowKey)
        for x, bit := range getBits(knotHash) {
            var coord = Coord{x, y}
            grid[coord] = bit != 0
        }
    }
    return grid
}


func getBits(knotHash []byte) (bits []byte) {
    for _, knot := range knotHash {
        var mask byte = 1
        for i := 0; i < 8; i++ {
            var bit = knot & mask
            bits = append(bits, bit)
            mask = mask << 1
        }
    }
    return
}


func countUsed(grid map[Coord]bool) (count int) {
    for _, isUsed := range grid {
        if isUsed {
            count++
        }
    }
    return
}


func main() {
    var keyString = os.Args[1]
    var grid = getDiskState(keyString)
    var numUsed = countUsed(grid)
    fmt.Println(numUsed)
}
